---
title: "Differential Expression Analysis strategy comparison for Model-based analysis of isobarically labeled proteomic data."
author: "Piotr Prostko, Joris Van Houtven"
date: '`r format(Sys.time(), "%B %d, %Y,%H:%M")`'
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    theme: flatly
    code_folding: "hide"
editor_options: 
  chunk_output_type: console
params:
  input_data_p: 'input_data_msstatstmt.Rds'
  suffix_p: 'msstatstmt'
  load_outputdata_p: FALSE
  subsample_p: 0
---

```{r, setup, include=FALSE}
# default knitr options for all chunks
knitr::opts_chunk$set(
  message=FALSE,
  warning=FALSE,
  fig.width=12,
  fig.height=7
)
```

```{r}
library(caret)
library(lme4)
library(lmerTest)
library(ggplot2)
library(stringi)
library(gridExtra)
library(ggfortify)
library(dendextend)
library(psych)
library(kableExtra)
library(tidyverse)
library(dtplyr)
library(DEqMS)
source('other_functions.R')
source('plotting_functions.R')
```

```{r}

load_outputdata=FALSE
data.list <- readRDS(params$input_data_p)
dat.l <- data.list$dat.l # data in long format
# which proteins were spiked in?
spiked.proteins <- dat.l %>% distinct(Protein) %>% filter(stri_detect(Protein, fixed='ups')) %>% pull %>% as.character
tmp=dat.l %>% distinct(Protein) %>% pull %>% as.character
# protein subsampling
if (params$subsample_p>0 & params$subsample_p==floor(params$subsample_p) & params$subsample_p<=length(tmp)){
  sub.prot <- tmp[sample(1:length(tmp), size=params$subsample_p)]
  if (length(spiked.proteins)>0) sub.prot <- c(sub.prot,spiked.proteins)
  dat.l <- dat.l %>% filter(Protein %in% sub.prot)
}
```

```{r}
# specify names of component variants
variant.names <- c('LMM', 'DEqMS', 'ANOVA_protein', 'ANOVA_PSM')
n.comp.variants <- length(variant.names)
# pick reference condition for making plots / doing DEA
referenceCondition <- '0.5'
# specify colours corresponding to biological conditions
condition.color <- tribble(
  ~Condition, ~Color,
  "0.125", 'black',
  "0.5", 'blue',
  "0.667", 'green',
  "1", 'red' )
# create data frame with sample info (distinct Run,Channel, Sample, Condition, Color)
sample.info <- get_sample_info(dat.l, condition.color)
channelNames <- remove_factors(unique(sample.info$Channel))
```

# Unit scale component: log2 transformation of reporter ion intensities

```{r}
dat.unit.l <- dat.l %>% mutate(response=log2(intensity)) %>% select(-intensity)
```

# Summarization component: no summarization

```{r}
# no summarization 
dat.summ.l <- dat.unit.l
```

# Normalization component: linear mixed model

```{r}
if (!load_outputdata){
dat.norm.l <- dat.summ.l
LMM1 <- lmer(response ~ Mixture + Mixture:TechRepMixture + Mixture:TechRepMixture:Channel + (1|Protein)  + (1|Mixture:TechRepMixture:Peptide), data=dat.summ.l)
dat.norm.l$response <- residuals(LMM1) + fixef(LMM1)['(Intercept)']
}
```

# QC plots

```{r}
if (!load_outputdata){
dat.nonnorm.summ.l <- aggFunc(dat.summ.l, 'response', group.vars=c('Mixture', 'TechRepMixture', 'Run', 'Channel', 'Condition', 'BioReplicate', 'Protein', 'Peptide'), 'median') 
dat.nonnorm.summ.l <- aggFunc(dat.nonnorm.summ.l, 'response', group.vars=c('Mixture', 'TechRepMixture', 'Run', 'Channel', 'Condition', 'BioReplicate', 'Protein'), 'median')

dat.norm.summ.l <- aggFunc(dat.norm.l, 'response', group.vars=c('Mixture', 'TechRepMixture', 'Run', 'Channel', 'Condition', 'BioReplicate', 'Protein', 'Peptide'), 'median')
dat.norm.summ.l <- aggFunc(dat.norm.summ.l, 'response', group.vars=c('Mixture', 'TechRepMixture', 'Run', 'Channel', 'Condition', 'BioReplicate', 'Protein'), 'median') 

# make data completely wide (also across runs)
## normalized data
dat.norm.summ.w2 <- pivot_wider(data=dat.norm.summ.l, id_cols=Protein, names_from=Run:Channel, values_from=response, names_sep=':')

## non-normalized data
dat.nonnorm.summ.w2 <- pivot_wider(data=dat.nonnorm.summ.l, id_cols=Protein, names_from=Run:Channel, values_from=response, names_sep=':') 
}
```

```{r, echo=FALSE}
if (load_outputdata) load(paste0('modelbased_DEA_outdata_', params$suffix_p, '.rda'))
```

## Boxplot

```{r}
par(mfrow=c(1,2))
boxplot_ils(dat.nonnorm.summ.l, 'raw')
boxplot_ils(dat.norm.summ.l, 'normalized')
```

## MA plot

MA plots of two single samples taken from condition 1 and condition 0.125, measured in different MS runs (samples *Mixture2_1:127C* and *Mixture1_2:129N*, respectively).

```{r}
p1 <- maplot_ils(dat.nonnorm.summ.w2, 'Mixture2_1:127C', 'Mixture1_2:129N', scale='log', 'raw', spiked.proteins)
p2 <- maplot_ils(dat.norm.summ.w2, 'Mixture2_1:127C', 'Mixture1_2:129N', scale='log', 'normalized', spiked.proteins)
grid.arrange(p1, p2, ncol=2)  
```

MA plots of all samples from condition `1` and condition `0.125` (quantification values averaged within condition).

```{r}
channels.num <- sample.info %>% filter(Condition=='1') %>% distinct(Sample) %>% pull
channels.denom <- sample.info %>% filter(Condition=='0.125') %>% distinct(Sample) %>% pull
p1 <- maplot_ils(dat.nonnorm.summ.w2, channels.num, channels.denom, scale='log', 'raw', spiked.proteins)
p2 <- maplot_ils(dat.norm.summ.w2, channels.num, channels.denom, scale='log', 'normalized', spiked.proteins)
grid.arrange(p1, p2, ncol=2)  
```

## PCA plot

### Using all proteins
```{r}
par(mfrow=c(1, 2))
pcaplot_ils(dat.nonnorm.summ.w2 %>% select(-'Protein'), info=sample.info, 'raw') 
pcaplot_ils(dat.norm.summ.w2 %>% select(-'Protein'), info=sample.info, 'normalized')
```

### Using spiked proteins only

```{r}
par(mfrow=c(1, 2))
pcaplot_ils(dat.nonnorm.summ.w2 %>% filter(Protein %in% spiked.proteins) %>% select(-'Protein'), info=sample.info, 'raw')
pcaplot_ils(dat.norm.summ.w2 %>% filter(Protein %in% spiked.proteins) %>% select(-'Protein'), info=sample.info, 'normalized')
```

## HC (hierarchical clustering) plot

### Using all proteins

```{r}
par(mfrow=c(1, 2))
dendrogram_ils(dat.nonnorm.summ.w2 %>% select(-'Protein'), info=sample.info, 'raw')
dendrogram_ils(dat.norm.summ.w2 %>% select(-'Protein'), info=sample.info, 'normalized')
```

## Run effect p-value plot

Below the histograms of p-values from the linear model explaining the response variable with Run as a covariate.

```{r}
dat <- list(dat.nonnorm.summ.l,dat.norm.summ.l)
names(dat) <- c('raw','normalized')
run_effect_plot(dat)
```

# DEA component

```{r, echo=FALSE}
# count how many peptides and PSMs per protein
PSMstat=dat.norm.l %>% group_by(Protein, Run, Channel) %>% summarise(PSMcount=n()) %>% group_by(Protein) %>% summarise(PSMcountMin=min(PSMcount), PSMcountMax=max(PSMcount))
#PSMstat %>% filter(PSMcountMin>1) %>% nrow

Peptidestat=dat.norm.l %>% group_by(Protein, Run, Channel) %>% summarise(Peptidecount=n_distinct(Peptide)) %>% group_by(Protein) %>% summarise(PeptidecountMin=min(Peptidecount), PeptidecountMax=max(Peptidecount))
```

## linear mixed model

Intra-protein correlation modelled with 1|Run:Channel random effect.

```{r}
if (!load_outputdata){
dat.dea <- emptyList(variant.names)
dat.dea$LMM <- lmm_dea(dat=dat.norm.l, mod.formula='response ~ Condition + (1|Run:Channel)', referenceCondition, scale='log')
}

# character vectors containing logFC and p-values columns
dea.cols <- colnames(dat.dea$LMM)
logFC.cols <- dea.cols[stri_detect_fixed(dea.cols, 'logFC')]
significance.cols <- dea.cols[stri_detect_fixed(dea.cols, 'q.mod')]
n.contrasts <- length(logFC.cols)
```

## DEqMS

```{r}
if (!load_outputdata){
# first compute PSM count per protein within run & channel, then for each protein separately take the minimum of the PSM counts (as in DEqMS vignette)
PSMcounts.df <- dat.norm.l %>% ungroup %>% group_by(Run, Channel, Protein) %>% summarize(PSMcount=n()) %>% group_by(Protein) %>% summarize(PSMcount=min(PSMcount,na.rm=TRUE)) %>% data.frame

design.matrix <- get_design_matrix(referenceCondition, sample.info)
d <- column_to_rownames(as.data.frame(dat.norm.summ.w2), 'Protein')
dat.dea$DEqMS <- deqms_test(dat=d, design.matrix, scale='log', PSMcounts.df)
}
```

## One-way ANOVA (protein level data)

```{r}
if (!load_outputdata){
dat.dea$ANOVA_protein <- get_anova(d, design.matrix, scale='log') 
}
```

## One-way ANOVA (PSM level data)
```{r}
if (!load_outputdata){
dat.dea$ANOVA_PSM <- lm_dea(dat=dat.norm.l, mod.formula='response ~ Condition', referenceCondition, scale='log') 
}
```

```{r, echo=FALSE}
# save output data
if (!load_outputdata){
save(dat.nonnorm.summ.l
     ,dat.norm.summ.l
     ,dat.nonnorm.summ.w2
     ,dat.norm.summ.w2
     ,dat.norm.l
     ,dat.summ.l
     ,dat.dea, file=paste0('modelbased_DEA_outdata_', params$suffix_p, '.rda'))
}
```

# Results comparison

## Confusion matrix

```{r, results='asis'}
cm <- conf_mat(dat.dea, 'q.mod', 0.05, spiked.proteins) 
print_conf_mat(cm, referenceCondition)
```

## Scatter plot

```{r}
scatterplot_ils(dat.dea, significance.cols, 'q-values', spiked.proteins, referenceCondition)
scatterplot_ils(dat.dea, logFC.cols, 'log2FC', spiked.proteins, referenceCondition)
```

## Volcano plot

```{r}
for (i in 1:n.contrasts){
  volcanoplot_ils(dat.dea, i, spiked.proteins, referenceCondition)
}
```

## Violin plot

Let's see whether the spiked protein fold changes make sense
```{r}
# plot theoretical value (horizontal lines) and violin per variant
violinplot_ils(lapply(dat.dea, function(x) x[spiked.proteins, logFC.cols]), referenceCondition)
```

# Conclusions

# Session information

```{r}
sessionInfo()
```